File: labs/packages/python-sdk/igris/models/ml.py
"""
Machine Learning models for Igris-engine SDK
"""

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union
from datetime import datetime
from enum import Enum

from .common import JobStatus


class MLTaskType(str, Enum):
    """Machine learning task types."""
    
    CLASSIFICATION = "classification"
    REGRESSION = "regression"
    CLUSTERING = "clustering"
    ANOMALY_DETECTION = "anomaly_detection"
    TIME_SERIES_FORECASTING = "time_series_forecasting"
    RECOMMENDATION = "recommendation"
    NATURAL_LANGUAGE_PROCESSING = "nlp"
    COMPUTER_VISION = "computer_vision"


class ModelType(str, Enum):
    """Supported model types."""
    
    LINEAR_REGRESSION = "linear_regression"
    LOGISTIC_REGRESSION = "logistic_regression"
    RANDOM_FOREST = "random_forest"
    GRADIENT_BOOSTING = "gradient_boosting"
    SVM = "svm"
    NEURAL_NETWORK = "neural_network"
    DEEP_LEARNING = "deep_learning"
    ENSEMBLE = "ensemble"


class TrainingStatus(str, Enum):
    """Training job status."""
    
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


@dataclass
class MLPipelineConfig:
    """Machine learning pipeline configuration."""
    
    name: str
    task_type: MLTaskType
    model_type: ModelType
    target_column: str
    feature_columns: List[str] = field(default_factory=list)
    training_data_path: Optional[str] = None
    validation_split: float = 0.2
    test_split: float = 0.1
    hyperparameters: Dict[str, Any] = field(default_factory=dict)
    preprocessing_steps: List[Dict[str, Any]] = field(default_factory=list)
    evaluation_metrics: List[str] = field(default_factory=list)
    cross_validation_folds: int = 5
    auto_feature_selection: bool = True
    auto_hyperparameter_tuning: bool = True
    early_stopping: bool = True
    description: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for API request."""
        return {
            "name": self.name,
            "task_type": self.task_type.value,
            "model_type": self.model_type.value,
            "target_column": self.target_column,
            "feature_columns": self.feature_columns,
            "training_data_path": self.training_data_path,
            "validation_split": self.validation_split,
            "test_split": self.test_split,
            "hyperparameters": self.hyperparameters,
            "preprocessing_steps": self.preprocessing_steps,
            "evaluation_metrics": self.evaluation_metrics,
            "cross_validation_folds": self.cross_validation_folds,
            "auto_feature_selection": self.auto_feature_selection,
            "auto_hyperparameter_tuning": self.auto_hyperparameter_tuning,
            "early_stopping": self.early_stopping,
            "description": self.description,
            "tags": self.tags
        }


@dataclass
class ModelMetrics:
    """Model evaluation metrics."""
    
    accuracy: Optional[float] = None
    precision: Optional[float] = None
    recall: Optional[float] = None
    f1_score: Optional[float] = None
    auc_roc: Optional[float] = None
    mse: Optional[float] = None
    rmse: Optional[float] = None
    mae: Optional[float] = None
    r2_score: Optional[float] = None
    confusion_matrix: Optional[List[List[int]]] = None
    classification_report: Optional[Dict[str, Any]] = None
    feature_importance: Optional[Dict[str, float]] = None
    custom_metrics: Dict[str, float] = field(default_factory=dict)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "accuracy": self.accuracy,
            "precision": self.precision,
            "recall": self.recall,
            "f1_score": self.f1_score,
            "auc_roc": self.auc_roc,
            "mse": self.mse,
            "rmse": self.rmse,
            "mae": self.mae,
            "r2_score": self.r2_score,
            "confusion_matrix": self.confusion_matrix,
            "classification_report": self.classification_report,
            "feature_importance": self.feature_importance,
            "custom_metrics": self.custom_metrics
        }


@dataclass
class TrainingJob:
    """Machine learning training job."""
    
    job_id: str
    pipeline_id: str
    status: TrainingStatus
    progress_percentage: float = 0.0
    current_epoch: Optional[int] = None
    total_epochs: Optional[int] = None
    training_loss: Optional[float] = None
    validation_loss: Optional[float] = None
    best_score: Optional[float] = None
    training_time_seconds: float = 0.0
    estimated_time_remaining: Optional[float] = None
    model_path: Optional[str] = None
    metrics: Optional[ModelMetrics] = None
    logs: List[str] = field(default_factory=list)
    error_message: Optional[str] = None
    created_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    
    @property
    def is_completed(self) -> bool:
        """Check if training is completed."""
        return self.status in (TrainingStatus.COMPLETED, TrainingStatus.FAILED, TrainingStatus.CANCELLED)
    
    @property
    def is_running(self) -> bool:
        """Check if training is running."""
        return self.status in (TrainingStatus.PENDING, TrainingStatus.RUNNING)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "job_id": self.job_id,
            "pipeline_id": self.pipeline_id,
            "status": self.status.value,
            "progress_percentage": self.progress_percentage,
            "current_epoch": self.current_epoch,
            "total_epochs": self.total_epochs,
            "training_loss": self.training_loss,
            "validation_loss": self.validation_loss,
            "best_score": self.best_score,
            "training_time_seconds": self.training_time_seconds,
            "estimated_time_remaining": self.estimated_time_remaining,
            "model_path": self.model_path,
            "metrics": self.metrics.to_dict() if self.metrics else None,
            "logs": self.logs,
            "error_message": self.error_message,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None
        }


@dataclass
class MLPipelineResult:
    """Result of ML pipeline execution."""
    
    pipeline_id: str
    job_id: str
    status: TrainingStatus
    model_id: Optional[str] = None
    model_path: Optional[str] = None
    metrics: Optional[ModelMetrics] = None
    predictions: Optional[List[Any]] = None
    feature_importance: Optional[Dict[str, float]] = None
    hyperparameters_used: Dict[str, Any] = field(default_factory=dict)
    training_summary: Dict[str, Any] = field(default_factory=dict)
    artifacts: Dict[str, str] = field(default_factory=dict)  # artifact_name -> path
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "pipeline_id": self.pipeline_id,
            "job_id": self.job_id,
            "status": self.status.value,
            "model_id": self.model_id,
            "model_path": self.model_path,
            "metrics": self.metrics.to_dict() if self.metrics else None,
            "predictions": self.predictions,
            "feature_importance": self.feature_importance,
            "hyperparameters_used": self.hyperparameters_used,
            "training_summary": self.training_summary,
            "artifacts": self.artifacts
        }


@dataclass
class PredictionRequest:
    """Request for model prediction."""
    
    model_id: str
    input_data: Union[Dict[str, Any], List[Dict[str, Any]]]
    return_probabilities: bool = False
    explain_predictions: bool = False
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for API request."""
        return {
            "model_id": self.model_id,
            "input_data": self.input_data,
            "return_probabilities": self.return_probabilities,
            "explain_predictions": self.explain_predictions
        }


@dataclass
class PredictionResult:
    """Result of model prediction."""
    
    predictions: List[Any]
    probabilities: Optional[List[List[float]]] = None
    explanations: Optional[List[Dict[str, Any]]] = None
    model_version: Optional[str] = None
    prediction_time_ms: float = 0.0
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "predictions": self.predictions,
            "probabilities": self.probabilities,
            "explanations": self.explanations,
            "model_version": self.model_version,
            "prediction_time_ms": self.prediction_time_ms
        }


@dataclass
class ModelInfo:
    """Information about a trained model."""
    
    model_id: str
    name: str
    model_type: ModelType
    task_type: MLTaskType
    version: str
    status: str
    accuracy: Optional[float] = None
    training_date: Optional[datetime] = None
    last_used: Optional[datetime] = None
    size_bytes: Optional[int] = None
    feature_columns: List[str] = field(default_factory=list)
    target_column: Optional[str] = None
    metrics: Optional[ModelMetrics] = None
    tags: List[str] = field(default_factory=list)
    description: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "model_id": self.model_id,
            "name": self.name,
            "model_type": self.model_type.value,
            "task_type": self.task_type.value,
            "version": self.version,
            "status": self.status,
            "accuracy": self.accuracy,
            "training_date": self.training_date.isoformat() if self.training_date else None,
            "last_used": self.last_used.isoformat() if self.last_used else None,
            "size_bytes": self.size_bytes,
            "feature_columns": self.feature_columns,
            "target_column": self.target_column,
            "metrics": self.metrics.to_dict() if self.metrics else None,
            "tags": self.tags,
            "description": self.description
        }

File: labs/packages/javascript-sdk/src/types/ml.ts
/**
 * Machine Learning types for Igris-engine JavaScript SDK
 */

/**
 * Machine learning task types
 */
export enum MLTaskType {
  CLASSIFICATION = 'classification',
  REGRESSION = 'regression',
  CLUSTERING = 'clustering',
  ANOMALY_DETECTION = 'anomaly_detection',
  TIME_SERIES_FORECASTING = 'time_series_forecasting',
  RECOMMENDATION = 'recommendation',
  NATURAL_LANGUAGE_PROCESSING = 'nlp',
  COMPUTER_VISION = 'computer_vision'
}

/**
 * Supported model types
 */
export enum ModelType {
  LINEAR_REGRESSION = 'linear_regression',
  LOGISTIC_REGRESSION = 'logistic_regression',
  RANDOM_FOREST = 'random_forest',
  GRADIENT_BOOSTING = 'gradient_boosting',
  SVM = 'svm',
  NEURAL_NETWORK = 'neural_network',
  DEEP_LEARNING = 'deep_learning',
  ENSEMBLE = 'ensemble'
}

/**
 * Training job status
 */
export enum TrainingStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

/**
 * Machine learning pipeline configuration
 */
export interface MLPipelineConfig {
  name: string;
  task_type: MLTaskType;
  model_type: ModelType;
  target_column: string;
  feature_columns?: string[];
  training_data_path?: string;
  validation_split?: number;
  test_split?: number;
  hyperparameters?: Record<string, unknown>;
  preprocessing_steps?: Array<Record<string, unknown>>;
  evaluation_metrics?: string[];
  cross_validation_folds?: number;
  auto_feature_selection?: boolean;
  auto_hyperparameter_tuning?: boolean;
  early_stopping?: boolean;
  description?: string;
  tags?: string[];
}

/**
 * Model evaluation metrics
 */
export interface ModelMetrics {
  accuracy?: number;
  precision?: number;
  recall?: number;
  f1_score?: number;
  auc_roc?: number;
  mse?: number;
  rmse?: number;
  mae?: number;
  r2_score?: number;
  confusion_matrix?: number[][];
  classification_report?: Record<string, unknown>;
  feature_importance?: Record<string, number>;
  custom_metrics?: Record<string, number>;
}

/**
 * Machine learning training job
 */
export interface TrainingJob {
  job_id: string;
  pipeline_id: string;
  status: TrainingStatus;
  progress_percentage: number;
  current_epoch?: number;
  total_epochs?: number;
  training_loss?: number;
  validation_loss?: number;
  best_score?: number;
  training_time_seconds: number;
  estimated_time_remaining?: number;
  model_path?: string;
  metrics?: ModelMetrics;
  logs: string[];
  error_message?: string;
  created_at?: string;
  started_at?: string;
  completed_at?: string;
}

/**
 * Result of ML pipeline execution
 */
export interface MLPipelineResult {
  pipeline_id: string;
  job_id: string;
  status: TrainingStatus;
  model_id?: string;
  model_path?: string;
  metrics?: ModelMetrics;
  predictions?: unknown[];
  feature_importance?: Record<string, number>;
  hyperparameters_used?: Record<string, unknown>;
  training_summary?: Record<string, unknown>;
  artifacts?: Record<string, string>; // artifact_name -> path
}

/**
 * Request for model prediction
 */
export interface PredictionRequest {
  model_id: string;
  input_data: Record<string, unknown> | Array<Record<string, unknown>>;
  return_probabilities?: boolean;
  explain_predictions?: boolean;
}

/**
 * Result of model prediction
 */
export interface PredictionResult {
  predictions: unknown[];
  probabilities?: number[][];
  explanations?: Array<Record<string, unknown>>;
  model_version?: string;
  prediction_time_ms: number;
}

/**
 * Information about a trained model
 */
export interface ModelInfo {
  model_id: string;
  name: string;
  model_type: ModelType;
  task_type: MLTaskType;
  version: string;
  status: string;
  accuracy?: number;
  training_date?: string;
  last_used?: string;
  size_bytes?: number;
  feature_columns: string[];
  target_column?: string;
  metrics?: ModelMetrics;
  tags: string[];
  description?: string;
}

/**
 * Hyperparameter tuning configuration
 */
export interface HyperparameterConfig {
  method: 'grid_search' | 'random_search' | 'bayesian' | 'genetic';
  parameters: Record<string, ParameterRange>;
  max_trials?: number;
  timeout_minutes?: number;
  early_stopping?: boolean;
  scoring_metric: string;
}

/**
 * Parameter range for hyperparameter tuning
 */
export interface ParameterRange {
  type: 'int' | 'float' | 'categorical' | 'boolean';
  min?: number;
  max?: number;
  values?: unknown[];
  distribution?: 'uniform' | 'log_uniform' | 'normal' | 'log_normal';
}

/**
 * Feature engineering configuration
 */
export interface FeatureEngineering {
  auto_feature_selection: boolean;
  feature_selection_method?: 'correlation' | 'mutual_info' | 'chi2' | 'rfe';
  max_features?: number;
  scaling_method?: 'standard' | 'minmax' | 'robust' | 'quantile';
  encoding_method?: 'onehot' | 'label' | 'target' | 'binary';
  handle_missing?: 'drop' | 'mean' | 'median' | 'mode' | 'constant';
  polynomial_features?: boolean;
  interaction_features?: boolean;
  custom_transformations?: Array<Record<string, unknown>>;
}

/**
 * Cross-validation configuration
 */
export interface CrossValidationConfig {
  method: 'kfold' | 'stratified_kfold' | 'time_series' | 'group_kfold';
  folds: number;
  shuffle?: boolean;
  random_state?: number;
  group_column?: string;
}

/**
 * Model deployment configuration
 */
export interface DeploymentConfig {
  deployment_name: string;
  model_id: string;
  environment: 'staging' | 'production';
  instance_type?: string;
  min_instances?: number;
  max_instances?: number;
  auto_scaling?: boolean;
  monitoring_enabled?: boolean;
  logging_level?: 'debug' | 'info' | 'warning' | 'error';
}

/**
 * Batch prediction job
 */
export interface BatchPredictionJob {
  job_id: string;
  model_id: string;
  input_path: string;
  output_path: string;
  status: TrainingStatus;
  progress_percentage: number;
  total_records?: number;
  processed_records?: number;
  error_records?: number;
  created_at?: string;
  started_at?: string;
  completed_at?: string;
  error_message?: string;
}

/**
 * Model monitoring metrics
 */
export interface ModelMonitoring {
  model_id: string;
  deployment_id: string;
  data_drift_score?: number;
  prediction_drift_score?: number;
  performance_metrics?: ModelMetrics;
  prediction_latency_p95?: number;
  prediction_throughput?: number;
  error_rate?: number;
  last_updated?: string;
  alerts?: ModelAlert[];
}

/**
 * Model alert configuration
 */
export interface ModelAlert {
  alert_id: string;
  type: 'data_drift' | 'performance_degradation' | 'latency' | 'error_rate';
  threshold: number;
  current_value: number;
  severity: 'low' | 'medium' | 'high' | 'critical';
  message: string;
  created_at: string;
  resolved_at?: string;
}

/**
 * A/B testing configuration for models
 */
export interface ABTestConfig {
  test_name: string;
  control_model_id: string;
  treatment_model_id: string;
  traffic_split: number; // 0-100, percentage to treatment
  success_metric: string;
  minimum_sample_size: number;
  max_duration_days: number;
  significance_level?: number;
}

/**
 * Model explainability request
 */
export interface ExplainabilityRequest {
  model_id: string;
  input_data: Record<string, unknown>;
  method: 'shap' | 'lime' | 'permutation' | 'integrated_gradients';
  top_k_features?: number;
}

/**
 * Model explainability result
 */
export interface ExplainabilityResult {
  feature_importances: Record<string, number>;
  explanation_method: string;
  prediction_value: unknown;
  baseline_value?: unknown;
  confidence_score?: number;
  visual_explanation?: string; // Base64 encoded image or URL
}

/**
 * AutoML configuration
 */
export interface AutoMLConfig {
  task_type: MLTaskType;
  target_column: string;
  training_data_path: string;
  time_budget_minutes?: number;
  quality_vs_speed?: 'speed' | 'balanced' | 'quality';
  interpretability?: 'low' | 'medium' | 'high';
  include_algorithms?: ModelType[];
  exclude_algorithms?: ModelType[];
  custom_scoring?: string;
  ensemble?: boolean;
}

/**
 * Model registry entry
 */
export interface ModelRegistryEntry {
  model_id: string;
  name: string;
  version: string;
  stage: 'development' | 'staging' | 'production' | 'archived';
  description?: string;
  tags: Record<string, string>;
  metrics: ModelMetrics;
  artifacts: Record<string, string>;
  created_by: string;
  created_at: string;
  updated_at: string;
  lineage?: ModelLineage;
}

/**
 * Model lineage tracking
 */
export interface ModelLineage {
  parent_models?: string[];
  training_data_sources: string[];
  feature_engineering_pipeline?: string;
  hyperparameter_tuning_job?: string;
  evaluation_datasets: string[];
  deployment_history: Array<{
    environment: string;
    deployed_at: string;
    deployed_by: string;
    status: string;
  }>;
}

export default {};